src/main/kotlin/com/mineinabyss/looty/ecs/systems/ItemRecipeSystem.kt

```diff
@@ -46,4 +54,35 @@ class ItemRecipeSystem : RepeatingSystem(), Listener {
     fun PlayerJoinEvent.showRecipesOnJoin() {
         player.discoverRecipes(discoveredRecipes)
     }
+
+    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
+    fun PrepareItemCraftEvent.onCraftWithLootyItem() {
+        val player = view.player as? Player ?: return
+        if (inventory.matrix.any { it?.notAllowedInRecipe(player, AllowedCraftingEnum.CRAFTING_TABLE) == true })
+            inventory.result = null
+    }
+
+    //TODO Make allowedInRecipe not require a player as there is no UUID on player_instanced items
+    /*@EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
+    fun FurnaceBurnEvent.onBurnLootyItem() {
+        if (this.willConsumeFuel() && fuel.notAllowedInRecipe(checkAllowedIn = AllowedCraftingEnum.FURNACE)) isCancelled = true
+    }*/
+
+    //TODO Smithing Table is an "UpgradeRecipe" not a normal crafting one
+    // Thus the vanilla behaviour is to forcefully copy the NBT of the input over even if result is set to null
+    // https://github.com/PaperMC/Paper/issues/6669 Sounds like we need to specify a recipe for this
+    // but not sure how since input and addition needs to be specified
+    /*@EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
+    fun PrepareSmithingEvent.onSmithingLootyItem() {
+        val player = view.player as? Player ?: return
+        SmithingRecipeIngredients(inventory.inputEquipment!!.toSerializable(), inventory.inputMineral!!.toSerializable()).toRecipe()
+        if (inventory.inputEquipment?.notAllowedInRecipe(player, AllowedCraftingEnum.SMITHING_TABLE) == true)
+            inventory.result = null
+    }*/
+
+    private fun ItemStack.notAllowedInRecipe(player: Player? = null, checkAllowedIn: AllowedCraftingEnum): Boolean {
+        val item = (if (player != null) this.toGearyOrNull(player) else this.toGearyFromUUIDOrNull()) ?: return false
+        return (item.has<LootyType>() && item.has<AllowInVanillaRecipes>() &&
+                checkAllowedIn !in item.get<AllowInVanillaRecipes>()!!.allowedIn)
```

Comment:
Improve null safety, i'd write this something like `item.get<AllowInVanillaRecipes>()?.allowedIn?.contains(checkAlloewdIn) ?: false`

Reply:
had this at one point but whilst fucking arounf with smithing i changed it, ill return it sure